File: 1428-jump-game-iii/jump-game-iii.java
class Solution {
    public boolean canReach(int[] arr, int start) {
        if(arr[start]==0){
            return true;
        }
        boolean [] visit=new boolean[arr.length];
        return reach(arr,start,visit);
    }
    public boolean reach(int[] arr,int i,boolean[] visit){
        if(i<0 || i>=arr.length || visit[i]){
            return false;
        }

        if(arr[i]==0){
            return true;
        }
        visit[i]=true;
        boolean l=reach(arr,i-arr[i],visit);
        boolean r=reach(arr,i+arr[i],visit);
        return l || r;

    }
}

File: 1428-jump-game-iii/jump-game-iii.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        vector<bool> vis(arr.size(), false);
        queue<int> q;
        q.push(start);
        vis[start] = true;
        while (!q.empty()) {
            int idx = q.front();
            q.pop();
            if (arr[idx] == 0) {
                return true;
            }
            if (idx - arr[idx] >= 0 && !vis[idx - arr[idx]]) {
                vis[idx - arr[idx]] = true;
                q.push(idx - arr[idx]);
            }
            if (idx + arr[idx] < arr.size() && !vis[idx + arr[idx]]) {
                vis[idx + arr[idx]] = true;
                q.push(idx + arr[idx]);
            }
        }
        return false;
    }
};